# Tik-Tok-Channel-Tracker



In diesem Notebook wird eine Methode vorgestellt, um ausgewählte TikTok-Kanäle systematisch zu verfolgen und deren Aktivitäten visuell darzustellen. Ziel ist es, für eine Sammlung interessanter Kanäle sämtliche Veröffentlichungen, Interaktionen und Veränderungen im Zeitverlauf zu erfassen und übersichtlich aufzubereiten.

### Live-Kanalverfolgung
Die vorgestellte Vorgehensweise beinhaltet das Sammeln von Metadaten zu Videos der ausgewählten TikTok-Kanäle mittels automatisierter Skripte. Diese Metadaten umfassen Informationen wie Ansichten, Likes, Kommentare und weitere relevante Kennzahlen. Durch periodisches Ausführen des Sammelprozesses entsteht ein umfangreicher Datensatz, der die Aktivitäten der Kanäle zu verschiedenen Zeitpunkten dokumentiert.

______ 


#### Visuelle Darstellung und Zeitreihenanalyse

Mit Hilfe von Visualisierungstools werden die gesammelten Daten aufbereitet. Diagramme und Grafiken zeigen die Entwicklung der Aktivitätskennzahlen über die Zeit und ermöglichen so den Vergleich von Kanälen sowie das Erkennen von Trends und Mustern. Für jede Visualisierung sind detaillierte Informationen zu einzelnen Videoeinträgen abrufbar.

- Die zeitliche Entwicklung individueller Accounts analysiert.
- Metadaten aller Medieninhalte systematisch erfasst
- Strategien zur Langzeitüberwachung und Auswertung der Aktivitätsdaten



____________

### ausgewählte Kanäle

Diese Methode steht exemplarisch für größe Datensätze zur Auswertung.

#### 67 Tiktok-Kanäle

```text
https://www.tiktok.com/@gereonbollmann
https://www.tiktok.com/@carolin_bachmann_mdb
https://www.tiktok.com/@christinabaumafd
https://www.tiktok.com/@roger_beckamp
https://www.tiktok.com/@renebochmann
https://www.tiktok.com/@brandesdirk
https://www.tiktok.com/@hannesgnauck
https://www.tiktok.com/@gerrithuy
https://www.tiktok.com/@malte.kaufmann
https://www.tiktok.com/@mikemoncsek.de
https://www.tiktok.com/@bernd.schattner.mdb
https://www.tiktok.com/@eugen_schmidt.mdb
https://www.tiktok.com/@janwenzelschmidt
https://www.tiktok.com/@edgar.naujok
https://www.tiktok.com/@joachim.wundrak
https://www.tiktok.com/@kayuweziegler71
https://www.tiktok.com/@dr.rainer.rothfuss
https://www.tiktok.com/@tino.chrupalla.afd
https://www.tiktok.com/@drberndbaumann
https://www.tiktok.com/@brandner_afd
https://www.tiktok.com/@goetzfroemming
https://www.tiktok.com/@enricokomning
https://www.tiktok.com/@muenzenmaier
https://www.tiktok.com/@beatrixvonstorchneu
https://www.tiktok.com/@joernkoenig_mdb
https://www.tiktok.com/@stefan_keuter
https://www.tiktok.com/@karstenhilseafd
https://www.tiktok.com/@marcbernhard
https://www.tiktok.com/@haraldweyel
https://www.tiktok.com/@alice_weidel_afd
https://www.tiktok.com/@rene_springer
https://www.tiktok.com/@sichertdeutschland
https://www.tiktok.com/@janwenzelschmidt
https://www.tiktok.com/@manfred.schiller.afd
https://www.tiktok.com/@ulschzi2
https://www.tiktok.com/@bernd.schattner.mdb
https://www.tiktok.com/@frank.rinck
https://www.tiktok.com/@martin_e_renner
https://www.tiktok.com/@mr_afd
https://www.tiktok.com/@protschkasposition
https://www.tiktok.com/@juergenpohlafd
https://www.tiktok.com/@tmpeterka_mdb
https://www.tiktok.com/@gerold_otten
https://www.tiktok.com/@jan_nolte_bundestag
https://www.tiktok.com/@edgar.naujok
https://www.tiktok.com/@muenzenmaier
https://www.tiktok.com/@volkermuenz_mdb
https://www.tiktok.com/@dr.rainer.kraft
https://www.tiktok.com/@steffenkotre
https://www.tiktok.com/@joernkoenig_mdb
https://www.tiktok.com/@enricokomning
https://www.tiktok.com/@kleinwaechterafd
https://www.tiktok.com/@stefan_keuter
https://www.tiktok.com/@leiferikholm
https://www.tiktok.com/@karstenhilseafd
https://www.tiktok.com/@martin_hess_klartext
https://www.tiktok.com/@jochenhaug_mdb
https://www.tiktok.com/@kaygottschalk
https://www.tiktok.com/@hannesgnauck
https://www.tiktok.com/@albrecht_glaser_mdb
https://www.tiktok.com/@markusfrohnmaier
https://www.tiktok.com/@goetzfroemming
https://www.tiktok.com/@dietmar.friedhoff.mdb
https://www.tiktok.com/@peterfelser_mdb
https://www.tiktok.com/@dr_espendiller
https://www.tiktok.com/@thomas.ehrhorn
https://www.tiktok.com/@gottfriedcurio





_________________

### yt-dlp

**yt-dlp** ist ein Tool zum Herunterladen von Videos oder Metadaten von verschiedenen Websites


```bash
yt-dlp --flat-playlist --dump-single-json https://www.tiktok.com/@goetzfroemming | jq .
```

Das Kommando ruft die Metadaten der TikTok-Playlist des angegebenen Nutzers (**@goetzfroemming**) in komprimierter Form ab, ohne die eigentlichen Videos herunterzuladen (**--flat-playlist**) und gibt sie im JSON-Format aus (**--dump-single-json**). Mit **jq .** werden die JSON-Daten formatiert und übersichtlich angezeigt.

Es wird eine JSON-Datei zurückgegben die viele wertvolle Informationen zu den im Tiktok-Kanal gespeicherten Videos enthält. Wir wollen das haben.

So sieht ein Ausschnitt davon aus:


....

```json
    }
      ],
      "ie_key": "TikTok",
      "_type": "url",
      "url": "https://www.tiktok.com/@goetzfroemming/video/7118684775510019333",
      "__x_forwarded_for_ip": null
    },
    {
      "id": "7117161498392284422",
      "formats": null,
      "subtitles": null,
      "http_headers": {
        "Referer": "https://www.tiktok.com/@goetzfroemming/video/7117161498392284422"
      },
      "channel": "Dr. Götz Frömming",
      "channel_id": "MS4wLjABAAAAH9ePnlsV7vniDNqjkTQazED0ZAVMMCl5QjI62JrIAZnxNMB0FsnvV0Ds4yL1TJnE",
      "uploader": "goetzfroemming",
      "uploader_id": "7034985201290347526",
      "channel_url": "https://www.tiktok.com/@MS4wLjABAAAAH9ePnlsV7vniDNqjkTQazED0ZAVMMCl5QjI62JrIAZnxNMB0FsnvV0Ds4yL1TJnE",
      "uploader_url": "https://www.tiktok.com/@goetzfroemming",
      "track": "Originalton",
      "artists": [
        "Dr. Götz Frömming"
      ],
      "duration": 43,
      "title": "Dokumentation „TEURO TOTAL“ auf dem YouTube-Kanal der #afdimbundestag  🎞 #eurokrise #geldpolitik #benzinpreis #afd",
      "description": "Dokumentation „TEURO TOTAL“ auf dem YouTube-Kanal der #afdimbundestag  🎞 #eurokrise #geldpolitik #benzinpreis #afd",
      "timestamp": 1657093304,
      "view_count": 13800,
      "like_count": 816,
      "repost_count": 57,
      "comment_count": 46,
      "thumbnails": [
        {
          "id": "cover",
          "url": "https://p16-sign-va.tiktokcdn.com/obj/tos-maliva-p-0068/b4ada5a519c44b88a8d95ab664fdfcc8_1657093305?lk3s=81f88b70&x-expires=1736690400&x-signature=mBHEhc32X%2BcudmhCSasdr2f65o0%3D&shp=81f88b70&shcp=-",
          "preference": -1
        },
        {
          "id": "dynamicCover",
          "url": "https://p77-sign-va.tiktokcdn.com/obj/tos-maliva-p-0068/7ebfb14ab14049129b4efa641be62683_1657093305?lk3s=81f88b70&x-expires=1736690400&x-signature=Smx0bU9hVMW10lf14AKipWQVFac%3D&shp=81f88b70&shcp=-",
          "preference": -2
        },
        {
          "id": "originCover",
          "url": "https://p77-sign-va.tiktokcdn.com/obj/tos-maliva-p-0068/ab5aa394f3d94351b176dc919cff5d2c_1657093306?lk3s=81f88b70&x-expires=1736690400&x-signature=Y%2BhdEKlP2vHPbH7GifzkChEzx%2FA%3D&shp=81f88b70&shcp=-",
          "preference": -1
        }
      ],
      "ie_key": "TikTok",
      "_type": "url",
      "url": "https://www.tiktok.com/@goetzfroemming/video/7117161498392284422",
      "__x_forwarded_for_ip": null
    },


    .....


    ...
    ..

    .


________________

Dieses Skript soll die Social-Media-Aktivitäten politischer TikTok-Konten überwachen. Es dient zur Sammlung und Speicherung von Metadaten aus TikTok-Videos, die von vorher ausgewählten politischen Kanälen gepostet werden. 
Zunächst wird eine SQLite-Datenbank eingerichtet, in der alle Informationen systematisch gespeichert werden. 
Für die Extraktion der Metadaten wird das Tool `yt-dlp` verwendet, um detaillierte Informationen über die Videos zu sammeln.
Das Skript ist so konzipiert, dass es eine Liste von TikTok-Kanal-Links aus einer Datei einliest, die die URLs der zu überwachenden politischen Kanäle enthält. Es verarbeitet diese Links nacheinander und stellt sicher, dass alle relevanten Daten extrahiert und gespeichert werden. 


In [1]:
import os
import sqlite3
import subprocess
import json
import logging

# Logging konfigurieren
logging.basicConfig(
    filename='metadata_extraction.log',
    level=logging.DEBUG,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logging.info("Skript gestartet")

# SQLite-Datenbank erstellen oder öffnen
def init_db(db_name="media_metadata.db"):
    logging.debug(f"Initialisiere Datenbank: {db_name}")
    conn = sqlite3.connect(db_name)
    cursor = conn.cursor()
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS media_metadata (
            id TEXT PRIMARY KEY,
            url TEXT,
            title TEXT,
            description TEXT,
            duration INTEGER,
            view_count INTEGER,
            like_count INTEGER,
            repost_count INTEGER,
            comment_count INTEGER,
            uploader TEXT,
            uploader_id TEXT,
            channel TEXT,
            channel_id TEXT,
            channel_url TEXT,
            track TEXT,
            album TEXT,
            artists TEXT,
            timestamp INTEGER,
            extractor TEXT
        )
    ''')
    conn.commit()
    logging.debug("Datenbank initialisiert")
    return conn

# Metadaten extrahieren mit yt-dlp
def extract_metadata(url):
    logging.debug(f"Extrahiere Metadaten für URL: {url}")
    try:
        result = subprocess.run(
            ["yt-dlp", "--flat-playlist", "--dump-single-json", url],
            capture_output=True,
            text=True
        )
        if result.returncode == 0:
            logging.debug(f"Metadaten erfolgreich extrahiert für URL: {url}")
            return json.loads(result.stdout)
        else:
            logging.error(f"Fehler beim Abrufen von {url}: {result.stderr}")
            return None
    except Exception as e:
        logging.error(f"Fehler beim Abrufen von {url}: {e}")
        return None

# Einzelne Video-Metadaten in SQLite speichern
def save_video_metadata(conn, video):
    cursor = conn.cursor()
    try:
        cursor.execute('''
            INSERT OR REPLACE INTO media_metadata (
                id, url, title, description, duration, view_count,
                like_count, repost_count, comment_count, uploader,
                uploader_id, channel, channel_id, channel_url, track,
                album, artists, timestamp, extractor
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ''', (
            video.get("id"),
            video.get("url"),
            video.get("title"),
            video.get("description"),
            video.get("duration"),
            video.get("view_count"),
            video.get("like_count"),
            video.get("repost_count"),
            video.get("comment_count"),
            video.get("uploader"),
            video.get("uploader_id"),
            video.get("channel"),
            video.get("channel_id"),
            video.get("channel_url"),
            video.get("track"),
            video.get("album"),
            ", ".join(video.get("artists", [])),
            video.get("timestamp"),
            video.get("extractor")
        ))
        conn.commit()
        logging.debug(f"Metadaten für Video ID={video.get('id')} gespeichert")
    except sqlite3.Error as e:
        logging.error(f"SQLite-Fehler bei Video ID={video.get('id')}: {e}")

# Playlist verarbeiten und Metadaten speichern
def process_playlist_metadata(conn, playlist_metadata):
    if not playlist_metadata or "entries" not in playlist_metadata:
        logging.debug("Keine Videos in der Playlist gefunden")
        return

    for video in playlist_metadata["entries"]:
        save_video_metadata(conn, video)

# Links aus Datei lesen und verarbeiten
def process_links(file_name="links.txt", db_name="media_metadata.db"):
    logging.debug(f"Beginne Verarbeitung von Links aus Datei: {file_name}")
    conn = init_db(db_name)
    if not os.path.exists(file_name):
        logging.error(f"Datei {file_name} nicht gefunden")
        return

    with open(file_name, "r") as file:
        links = [line.strip() for line in file if line.strip()]

    logging.debug(f"Gefundene Links: {links}")

    for url in links:
        logging.info(f"Verarbeite URL: {url}")
        playlist_metadata = extract_metadata(url)
        if playlist_metadata:
            process_playlist_metadata(conn, playlist_metadata)
        else:
            logging.warning(f"Keine Metadaten gefunden für URL: {url}")

    conn.close()
    logging.debug("Verarbeitung abgeschlossen")

if __name__ == "__main__":
    process_links()


____________

### Datenbank sichten

Die sqlite3 Datenbank wird zur weiteren Verarbeitung in ein dataframe gespeichert.

In [2]:
# Datenbankinhalt in einen Pandas-DataFrame laden

import pandas as pd
import sqlite3
def db_to_dataframe(db_name="media_metadata.db"):
    conn = sqlite3.connect(db_name)
    query = "SELECT * FROM media_metadata"
    try:
        df = pd.read_sql_query(query, conn)
        return df
    except sqlite3.Error as e:
        print(f"Fehler: {e}")
        return None
    finally:
        conn.close()

df = db_to_dataframe()

df

id                                                url  \
0          1  https://www.tiktok.com/@roger_beckamp/video/68...   
1          2  https://www.tiktok.com/@roger_beckamp/video/68...   
2          3  https://www.tiktok.com/@roger_beckamp/video/68...   
3          4  https://www.tiktok.com/@roger_beckamp/video/68...   
4          5  https://www.tiktok.com/@roger_beckamp/video/68...   
...      ...                                                ...   
10139  10140  https://www.tiktok.com/@kleinwaechterafd/video...   
10140  10141  https://www.tiktok.com/@kleinwaechterafd/video...   
10141  10142  https://www.tiktok.com/@protschkasposition/vid...   
10142  10143  https://www.tiktok.com/@protschkasposition/vid...   
10143  10144  https://www.tiktok.com/@goetzfroemming/video/7...   

                                                   title  \
0                                                 #peace   
1                                                          
2                                                          
3                                                          
4                                                          
...                                                  ...   
10139  Typisches EU-Gaga: 150 #EU-Beamte űberwachen d...   
10140  Keine #Zensur im #Netz! #Freiheit ist unser hö...   
10141          #AfD #Deutschland #fürdich #Weidel #Musk    
10142  #AfD #Deutschland #fürdich #Weidel #Musk #Prot...   
10143  Wussten wir auf Deutsch schon immer, aber tut ...   

                                             description  duration  \
0                                                 #peace      47.0   
1                                                             38.0   
2                                                              8.0   
3                                                             36.0   
4                                                             25.0   
...                                                  ...       ...   
10139  Typisches EU-Gaga: 150 #EU-Beamte űberwachen d...      15.0   
10140  Keine #Zensur im #Netz! #Freiheit ist unser hö...     231.0   
10141          #AfD #Deutschland #fürdich #Weidel #Musk       48.0   
10142  #AfD #Deutschland #fürdich #Weidel #Musk #Prot...      60.0   
10143  Wussten wir auf Deutsch schon immer, aber tut ...      24.0   

       view_count  like_count  repost_count  comment_count  \
0          882900       58700          8271           1212   
1           46900         771            51             13   
2          333700        2343            38             43   
3          276800       10900           371             93   
4          125200        3605           257             57   
...           ...         ...           ...            ...   
10139        1631         252             4             25   
10140        2027         396            20             30   
10141         274          32             0              4   
10142         287          56             0              3   
10143         280          28             0              2   

                 uploader          uploader_id                   channel  \
0           roger_beckamp  6847547576624923654             Roger Beckamp   
1           roger_beckamp  6847547576624923654             Roger Beckamp   
2           roger_beckamp  6847547576624923654             Roger Beckamp   
3           roger_beckamp  6847547576624923654             Roger Beckamp   
4           roger_beckamp  6847547576624923654             Roger Beckamp   
...                   ...                  ...                       ...   
10139    kleinwaechterafd  7025673845487289350  Norbert Kleinwächter AfD   
10140    kleinwaechterafd  7025673845487289350  Norbert Kleinwächter AfD   
10141  protschkasposition  7030758767387444230   Stephan Protschka (MdB)   
10142  protschkasposition  7030758767387444230   Stephan Protschka (MdB)   
10143      goetzfroemming  7034985201290347526         Dr. Göt

## 67 Kanäle und 10144 Videos


### How much is the fish?
#### Wer hat die meistern Likes,Kommentare,Repost oder Views ?

Das Skript vergleicht verschiedene TikTok-Kanäle anhand der Metriken "view_count", "like_count", "repost_count" und "comment_count". Es summiert die Werte jeder Metrik für jeden Kanal und erstellt für jede Metrik ein eigenes Balkendiagramm. Die Diagramme zeigen, wie die einzelnen Kanäle in Bezug auf diese Interaktionen abschneiden. 

In [12]:
import altair as alt
import pandas as pd

# Summe für jede Metrik pro Channel berechnen
columns_to_plot = ["view_count", "like_count", "repost_count", "comment_count"]
aggregated_data = (
    df.groupby("channel")[columns_to_plot].sum().reset_index()
    .melt(id_vars="channel", var_name="metric", value_name="value")
)

# Interaktive Balkendiagramme erstellen
charts = []
for metric in columns_to_plot:
    chart = (
        alt.Chart(aggregated_data[aggregated_data["metric"] == metric])
        .mark_bar()
        .encode(
            x=alt.X("channel:O", title="Channel", sort="-y"),
            y=alt.Y("value:Q", title=f"Total {metric.capitalize()}"),
            color=alt.Color("channel:N", legend=None),
            tooltip=[
                alt.Tooltip("channel:N", title="Channel"),
                alt.Tooltip("value:Q", title=f"Total {metric.capitalize()}"),
            ],
        )
        .properties(
            title=f"Total {metric.capitalize()} by Channel",
            height=300,  # Höhe für besseres Layout
            width=1200    # Breite für bessere Darstellung
        )
    )
    # Trennelement: Linie unter dem Chart
    separator = (
        alt.Chart(pd.DataFrame({"y": [0]}))
        .mark_rule(strokeDash=[5, 5], size=1)
        .encode(y=alt.Y("y:Q", axis=None))
        .properties(height=10, width=800)
    )
    charts.extend([chart, separator])

# Subplots mit Trennelementen darstellen
final_chart = alt.vconcat(*charts).resolve_scale(color="independent")

# Konfiguration auf das gesamte Diagramm anwenden
final_chart = final_chart.configure_axis(
    labelFontSize=12,  # Schriftgröße der Achsenbeschriftung
    titleFontSize=16   # Schriftgröße des Achsentitels
)

final_chart


alt.VConcatChart(...)

In [54]:
import pandas as pd
import altair as alt
from IPython.display import display, HTML

# Annahme: df ist bereits definiert und enthält die Daten

columns_to_plot = ["view_count", "like_count", "repost_count", "comment_count"]
top_n = 50
charts = []

for col in columns_to_plot:
    # Top 50 Einträge für die jeweilige Spalte auswählen und absteigend sortieren
    df_sorted = df.nlargest(top_n, col).sort_values(by=col, ascending=False).copy()
    # Eindeutige Labels erstellen, um Mehrfachvorkommen desselben Kanals zu unterscheiden
    df_sorted["channel_label"] = df_sorted["channel"] + " (ID:" + df_sorted["id"].astype(str) + ")"
    
    chart = (
        alt.Chart(df_sorted)
        .mark_bar()
        .encode(
            x=alt.X(f"{col}:Q", title=col.capitalize(), axis=alt.Axis(format=".0f")),
            # Y-Achse sortiert nach dem aktuellen DataFrame, der absteigend sortiert ist
            y=alt.Y("channel_label:N", title="Channel", sort=df_sorted["channel_label"].tolist()),
            color=alt.Color("channel:N", title="Channel"),
            tooltip=[
                "url", 
                "title", 
                "duration", 
                "view_count", 
                "like_count", 
                "repost_count", 
                "comment_count"
            ],
            href="url:N"  # Beim Klicken auf den Balken wird die URL geöffnet
        )
        .properties(
            title=f"Top {top_n} - {col.capitalize()}",
            width=800,
            height=600,
        )
    )
    charts.append(chart)

# Alle Subplots vertikal anordnen
final_chart = alt.vconcat(*charts)
final_chart.show()

# Tabellen für jede Metrik anzeigen
for col in columns_to_plot:
    # Top 50 Einträge erneut sortieren und vorbereiten
    df_sorted = df.nlargest(top_n, col).sort_values(by=col, ascending=False).copy()
    df_sorted["channel_label"] = df_sorted["channel"] + " (ID:" + df_sorted["id"].astype(str) + ")"
    
    # Überschrift für die Tabelle
    display(HTML(f"<h3>Top {top_n} - {col.capitalize()}</h3>"))
    # Schöne Tabelle mit Styling anzeigen und Index ausblenden
    styled_table = df_sorted.style.set_table_attributes('class="table table-striped"').hide(axis='index')
    display(styled_table)


alt.VConcatChart(...)

_____________

### Veröffentlichungen pro Tag 

Das Skript verarbeitet Daten zu Videoveröffentlichungen, wandelt Zeitstempel in jahres- und tagesbezogene Formate um und gruppiert sie nach Jahr, Tag und Kanal, um die Anzahl der Veröffentlichungen zu zählen. Anschließend erstellt es ein facettiertes Balkendiagramm, das die tägliche Veröffentlichungsaktivität pro Jahr und Kanal visualisiert.

Unter den Ausgewählten Kanälen lassen sich somit viele Trends in der Nutzung abbilden.

In [30]:
import pandas as pd
import altair as alt

# Zeilenbegrenzung deaktivieren
alt.data_transformers.disable_max_rows()

# Zeitbezogene Spalten erstellen
df["datetime"] = pd.to_datetime(df["timestamp"], unit='s')
df["year"] = df["datetime"].dt.year
df["day"] = df["datetime"].dt.floor('D')

# Gruppierung: Anzahl der Videos pro Tag und Kanal zählen
df_counts = (
    df.groupby(["year", "day", "channel"])
      .size()
      .reset_index(name='count')
)

# Sicherstellen, dass 'day' im richtigen Format vorliegt
df_counts["day"] = pd.to_datetime(df_counts["day"])

# Altair-Chart erstellen: Subplot für jedes Jahr mit jahresspezifischer x-Achse
chart = alt.Chart(df_counts).mark_bar().encode(
    x=alt.X("day:T", 
            title="Datum", 
            scale=alt.Scale(nice=False)  # Zeigt die spezifischen Tage jedes Jahres
           ),
    y=alt.Y("sum(count):Q", title="Anzahl Veröffentlichungen"),
    color=alt.Color("channel:N", title="Kanal", legend=alt.Legend(orient="right", labelFontSize=12, titleFontSize=14)),
    tooltip=[
        alt.Tooltip("day:T", title="Datum"),
        alt.Tooltip("channel:N", title="Kanal"),
        alt.Tooltip("sum(count):Q", title="Veröffentlichungen")
    ]
).properties(
    width=800  # Breiterer Plot
).facet(
    row=alt.Row("year:N", title="Jahr")
).resolve_scale(
    x="independent",  # Unabhängige x-Achse für jedes Jahr
    y="independent"
)

chart


alt.FacetChart(...)